src/rust/engine/src/rule_graph.rs

```diff
@@ -0,0 +1,551 @@
+            was_unfulfillable = True
+            continue
+
+          add_rules_to_graph(entry,
+                             (selector, selector.input_product_selector),
+                             initial_rules_or_literals)
+          add_rules_to_graph(entry,
+                             (selector, selector.projected_product_selector),
+                             tuple(rules_for_dependencies))
+                             */
+
+              },
+              &Selector::SelectProjection(ref select) =>{
+                let initial_selector = select.input_product;
+                let initial_rules_or_literals = rhs_for_select(&self.tasks,
+                                                               entry.subject_type(),
+                                                               &Select { product: initial_selector, variant_key: None });
+                if initial_rules_or_literals.is_empty() {
+                  mark_unfulfillable(&mut unfulfillable_rules,
+                                     &entry,
+                                     entry.subject_type(),
+                                     format!("no matches for {:?} when resolving {:?}", initial_selector, initial_selector)); // might be better as {} with display derived
+                  was_unfulfillable = true;
+                  continue
+                }
+                /*
+                      # TODO, could validate that input product has fields
+          initial_rules_or_literals = _find_rhs_for_select(entry.subject_type,
+                                                           selector.input_product_selector)
+          if not initial_rules_or_literals:
+            mark_unfulfillable(entry,
+                               entry.subject_type,
+                               'no matches for {} when resolving {}'
+                               .format(selector.input_product_selector, selector))
+            was_unfulfillable = True
+            continue
+
+          projected_rules = _find_rhs_for_select(selector.projected_subject,
+                                                 selector.projected_product_selector)
+          if not projected_rules:
+            mark_unfulfillable(entry,
+                               selector.projected_subject,
+                               'no matches for {} when resolving {}'
+                               .format(selector.projected_product_selector, selector))
+            was_unfulfillable = True
+            continue
+
+          add_rules_to_graph(entry,
+                             (selector, selector.input_product_selector),
+                             initial_rules_or_literals)
+          add_rules_to_graph(entry,
+                             (selector, selector.projected_product_selector),
+                             projected_rules)
+                */
+
+              },
+              &Selector::Task(ref select) =>{
+                // TODO, not sure what task is in this context exactly
```

Comment:
It's for symmetry: every `Node` type has a `Selector` currently. But probably possible to remove.